vowpalwabbit/interactions.cc

```diff
@@ -227,197 +290,101 @@ void sort_and_filter_duplicate_interactions(
-// for doublecheck.
-
-//#define DEBUG_EVAL_COUNT_OF_GEN_FT
-#ifdef DEBUG_EVAL_COUNT_OF_GEN_FT
-struct eval_gen_data
+std::vector<namespace_index> indices_to_values_one_based(
+    const std::vector<size_t>& indices, const std::set<namespace_index>& values)
 {
-  size_t& new_features_cnt;
-  float& new_features_value;
-  eval_gen_data(size_t& features_cnt, float& features_value)
-      : new_features_cnt(features_cnt), new_features_value(features_value)
+  std::vector<namespace_index> result;
+  result.reserve(indices.size());
+  for (size_t i = 0; i < indices.size(); i++)
   {
+    auto it = values.begin();
+    std::advance(it, indices[i] - 1);
+    result.push_back(*it);
   }
-};
-
-void ft_cnt(eval_gen_data& dat, const float fx, const uint64_t)
-{
-  ++dat.new_features_cnt;
-  dat.new_features_value += fx * fx;
+  return result;
 }
-#endif
-
-// lookup table of factorials up tu 21!
-constexpr int64_t fast_factorial[] = {1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880, 3628800, 39916800, 479001600,
-    6227020800, 87178291200, 1307674368000, 20922789888000, 355687428096000, 6402373705728000, 121645100408832000,
-    2432902008176640000};
-constexpr size_t size_fast_factorial = sizeof(fast_factorial) / sizeof(*fast_factorial);
 
-// helper factorial function that allows to perform:
-// n!/(n-k)! = (n-k+1)*(n-k+2)..*(n-1)*n
-// by specifying n-k as second argument
-// that helps to avoid size_t overflow for big n
-// leave second argument = 1 to get regular factorial function
-
-inline size_t factor(const size_t n, const size_t start_from = 1)
+std::vector<namespace_index> indices_to_values_ignore_last_index(
+    const std::vector<size_t>& indices, const std::set<namespace_index>& values)
 {
-  if (n <= 0) return 1;
-  if (start_from == 1 && n < size_fast_factorial) return static_cast<size_t>(fast_factorial[n]);
-
-  size_t res = 1;
-  for (size_t i = start_from + 1; i <= n; ++i) res *= i;
-  return res;
+  std::vector<namespace_index> result;
+  result.reserve(indices.size() - 1);
+  for (size_t i = 0; i < indices.size() - 1; i++)
+  {
+    auto it = values.begin();
+    std::advance(it, indices[i]);
+    result.push_back(*it);
+  }
+  return result;
 }
 
-// returns number of new features that will be generated for example and sum of their squared values
-void eval_count_of_generated_ft(bool permutations, const std::vector<std::vector<namespace_index>>& interactions,
-    const std::array<features, NUM_NAMESPACES>& feature_spaces, size_t& new_features_cnt, float& new_features_value)
+std::vector<std::vector<namespace_index>> generate_combinations_with_repetition(
+    const std::set<namespace_index>& namespaces, size_t num_to_pick)
 {
-  new_features_cnt = 0;
-  new_features_value = 0.;
-
-  v_array<float> results;
-
-  if (permutations)
+  std::vector<std::vector<namespace_index>> result;
+  // This computation involves factorials and so can only be done with relatively small inputs.
+  if ((namespaces.size() + num_to_pick) <= 21)
```

Comment:
why 21 in particular?

Reply:
I can add a comment but 21 is the largest factorial that fits in 64 bits, 22 would overflow.